cmd/kubectl-gadget/utils/trace.go

```diff
@@ -465,52 +468,77 @@ func getTraceWatcher(traceID string) (watch.Interface, error) {
 	// Get a watcher on all the traces which have the same ID.
 	// Indeed, all the traces on different nodes but linked to one gadget share
 	// the same ID.
-	watcher, err := getTraceWatcher(traceID)
+	// We will also begin to monitor events since the above GET of the traces
+	// list.
+	watcher, err := getTraceWatcher(traceID, traceList.ListMeta.ResourceVersion)
 	if err != nil {
 		return nil, err
 	}
 
 	nodeErrors := make(map[string]string)
 	nodeWarnings := make(map[string]string)
 
+	// We will only watch the traces which were not ready.
+	tracesToWatch := tracesNumber - tracesReady
+	watchedTracesNumber := 0
+
 	ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), traceTimeout)
 	_, err = untilWithoutRetry(ctx, watcher, func(event watch.Event) (bool, error) {
 		// This function will be executed until:
-		// 1. The number of watched traces equals the number of traces, i.e. we
-		// dealt with the traces which interest us.
+		// 1. The number of watched traces equals the number of traces to watch,
+		// i.e. we dealt with the traces which interest us.
 		// 2. Or it returns an error.
 		// 3. Or time out is fired.
 		// NOTE In case 2 and 3, it exists, at least, one trace we did not deal
 		// with.
 
-		// Deal particularly with error.
-		if event.Type == watch.Error {
+		switch event.Type {
+		case watch.Deleted:
+			// If for some strange reasons (e.g. users deleted a trace during this
+			// operation) a trace is deleted, we need to take care of this by
+			// decrementing the tracesNumber.
+			// Otherwise we would still wait for the old number and we would timeout.
+			tracesNumber--
+		case watch.Modified:
+			// We will deal with this type of event below
+		case watch.Error:
+			// Deal particularly with error.
 			return false, err
-		}
-
-		// We are only interested in Added and Modified event, as we want
-		// Status.State value to change.
-		// More particularly, we monitor Added event for gadget like dns and
-		// Modified for gadget like seccompadvisor.
-		if event.Type != watch.Added && event.Type != watch.Modified {
+		case watch.Added:
+			// createTraces() creates traces synchronously.
+			// So, if a watch.Added event occurs it means there is a problem (e.g. the
+			// user creates a trace by snooping on the traceID of existing traces).
+			return false, fmt.Errorf("No traces with the given traceID (%s) should be created", traceID)
+		default:
+			// We are not interested in other event types.
 			return false, nil
 		}
 
-		if event.Type == watch.Added {
-			// The API adds fake watch.Added events, so we use them to count the
-			// number of trace:
-			// To establish initial state, the watch begins with synthetic "Added"
-			// events of all resources instances that exist at the starting resource
-			// version
-			tracesNumber++
-		}
-
 		trace := event.Object.(*gadgetv1alpha1.Trace)
 
 		if trace.Status.OperationWarning != "" {
```

Comment:
> statement to explain why having a warning does not mean that trace can be considered as "watched" (i.e. watchedTracesNumber is not increased). 

`watchedTracesNumber` is actually increased there on L522. I just removed it to avoid counting it twice.

Reply:
Absolutely! I missed it, sorry 😅